lib/OpenQA/Worker.pm

```diff
@@ -129,28 +362,292 @@ sub prepare_cache_directory {
     return $shared_cache;
 }
 
+sub accept_job {
+    my ($self, $client, $job_info) = @_;
+
+    if ($self->current_job) {
+        die 'attempt to accept a new job although there is already a job running';
+    }
+
+    # instantiate new job
+    my $webui_host = $client->webui_host;
+    my $job        = OpenQA::Worker::Job->new($self, $client, $job_info);
+    $job->on(
+        status_changed => sub {
+            $self->_handle_job_status_changed(@_);
+        });
+
+    remove_log_channel('autoinst');
+    remove_log_channel('worker');
+    add_log_channel('autoinst', path => 'autoinst-log.txt', level => 'debug');
+    add_log_channel(
+        'worker',
+        path    => 'worker-log.txt',
+        level   => $self->settings->global_settings->{LOG_LEVEL} // 'info',
+        default => 'append',
+    );
+
+    $self->current_job($job);
+    $self->current_webui_host($webui_host);
+    $job->accept();
+}
+
+# stops the current job and (if there is one) and terminates the worker
 sub stop {
-    my ($reason) = @_;
+    my ($self, $reason) = @_;
 
-    if ($job) {
-        Mojo::IOLoop->next_tick(
-            sub {
-                stop_job($reason);
+    $self->{_shall_terminate} = 1;
+
+    my $current_job = $self->current_job;
+    if (!$current_job) {
+        # FIXME: better stop gracefully?
+        Mojo::IOLoop->stop;
+        return undef;
+    }
+
+    if ($current_job->status eq 'setup') {
+        # stop job directly during setup because the IO loop is blocked by isotovideo.pm during setup
```

Comment:
mentioning "isotovideo" in this context sounds wrong to me as we are not in the isotovideo-engine